### PR TITLE
Don't recommend setting html_root_url

### PR DESCRIPTION
**Pull request by jyn514**
See https://github.com/rust-lang/api-guidelines/issues/229 for discussion.

r? @BurntSushi

### CODE REVIEW

**Review by BurntSushi — APPROVED** [review]
LGTM, but I don't have merge permission on this repo. And it would probably be good to give it a few days for others to chime in.

**Comment by dtolnay** [comment]
FWIW in my work codebase, in which all Rust projects (and all other languages) are built by Buck, there is no Cargo involved and Buck makes `rustc` and `rustdoc` invocations directly. See https://buck.build/rule/rust_library.html. When building documentation of a Rust target (`buck build path/to/my/target#doc`) third party code without an html_root_url would not be able to be linked.

**Comment by jyn514** [comment]
@dtolnay I'm not very familiar with buck - does it use a shared build directory between dependencies? If so, rustdoc will pick up the documentation from the other dependencies without either `--extern-html-root-url` or `html_root_url`.

**Comment by dtolnay** [comment]
It does not use a shared build directory between dependencies. For example the output of `buck build path/to/my/target#doc` would always go into `buck-out/gen/path/to/my/target#doc`. If that target depends on some other target, the other target's docs are correspondingly somewhere else.

**Comment by jyn514** [comment]
Hmm, I see. I'd expect most of the links to be broken then :/ I don't know many libraries setting `html_root_url`.

On nightly there's `--extern-html-root-url`, but the docs say 1.33 is supported, so it wouldn't work on stable.

**Comment by dtolnay** [comment]
We're using the 1.48 prerelease so that's not a problem. We actually do use `--extern-html-root-url` for cross linking first-party code.

Regarding this:

> When rustdoc needs to generate a link to an item in an external crate, it will first check if the extern crate has been documented locally on-disk, and if so link directly to it. Failing that, it will use the URL given by the --extern-html-root-url command-line flag if available. If that is not available, then it will use the html_root_url value in the extern crate if it is available. If that is not available, then the extern items will not be linked.

The reason I don't think this fits our use case is that the priority order is backwards. We can easily make Buck always pass a docs.rs extern-html-root-url for all third-party crates (like you might have been expecting https://github.com/rust-lang/cargo/issues/8296 to wire up in the Cargo case) but that's wrong: Buck doesn't contain a Rust parser so it would have no idea whether the crate it's building already contains an `html_root_url` attribute. I think we'd be all set if `html_root_url` took precedence over extern-html-root-url but I'm not sure what other use case that breaks. Maybe this needs some separate flag.

**Comment by jyn514** [comment]
I'm trying to imagine the scenario where `html_root_url` should take precedence - are you imagining pointing to a site other than docs.rs, like https://api.rocket.rs/ or something? I guess it does make sense in that case.

In general I don't think there's a clear answer here: docs.rs *definitely* wants --extern-html-root-url to take precedence, so we can't break that use case either. I'd be willing to add a switch toggling the precedence though.

**Comment by dtolnay** [comment]
Yes exactly. Or third party code that's not in the crates.io / docs.rs ecosystem at all.

**Comment by jyn514** [comment]
Ok, so it sounds like there are two possible paths forward, both of which allow getting rid of `html_root_url` in the common case when you just want to link to docs.rs.

1. Have a flag for `html_root_url` to take precedence over `--extern-html-root-url`. Buck/Bazel/etc. can unconditionally pass `--extern-html-root-url` without generating broken links, and the documentation will link to the right place automatically. Docs.rs can unconditionally pass the flag because it knows all crates from crates.io are documented on docs.rs.
2. Have rustdoc link to docs.rs whenever `html_root_url` is unset and `--extern-html-root-url` is unset. Docs.rs continues passing `--extern-html-root-url` to override external docs (since it never wants to link to https://api.rocket.rs), buck doesn't have to pass it except when it actually wants to override docs.rs, and https://github.com/rust-lang/cargo/issues/8296 is no longer necessary.

I think doing both would be a mistake - do you have a preference for one or the other?

**Comment by jyn514** [comment]
cc @rust-lang/rustdoc, @rust-lang/cargo

**Comment by dtolnay** [comment]
Speaking for my own codebase the **absolute best** for us would actually be a rustc flag, not a rustdoc flag, which specified a "fallback html_root_url" for the *current crate* i.e. not DEP=URL like for --extern-html-root-url. We'd pass that from Buck during the compilation of third party code, and then when a doc build later depends on it, the link would go to the right place i.e. the crates actual html_root_url if it provides one, otherwise the docs.rs fallback fed by our Buck config.

**Comment by BurntSushi** [comment]
Either way, it seems like we don't need to suggest setting `html_root_url` in the API guidelines?

**Comment by jyn514** [comment]
@BurntSushi well, neither suggestion has been implemented yet. It depends whether you think the edge cases are worth worrying about.

**Comment by ehuss** [comment]
> 2. Have rustdoc link to docs.rs whenever `html_root_url` is unset and `--extern-html-root-url` is unset.

In this situation, how would rustdoc know this is a crates.io crate?  I don't think rustdoc should have any automatic linking to docs.rs without some kind of command-line opt-in (or `-Zrustdoc-map`, which knows which crates exist on crates.io).

> Either way, it seems like we don't need to suggest setting `html_root_url` in the API guidelines?

I think it would be reasonable to remove it.  There are few scenarios where it is useful.  One is running `cargo doc --no-deps`, but you want dependency links to still work. In that case, I think the user should just remove `--no-deps`, or use `-Zrustdoc-map`.  Only about 4% of crates on crates.io set it, of those only 225 set it to a non-docs.rs site.  In practice, I don't think it is in enough widespread use to warrant suggesting *everyone* set it.  I think it only makes sense if you really want to link to a non-docs.rs site, which very few do.

**Comment by jyn514** [comment]
> In this situation, how would rustdoc know this is a crates.io crate?

Ahh, good point, I hadn't thought about path dependencies or private registries. I think having an option for `html_root_url` to override `--extern-html-root-url` is the way to go, then.

**Comment by jyn514** [comment]
>  Should there be some way for the crate author to control the link destination? The documentation URL set in Cargo.toml is never consulted. I think this is unlikely to work with rustdoc, because it expects a certain structure to the API docs, and documentation URLs often point to user guides, or other things that aren't API docs. The #![doc(html_root_url = "...")] attribute is overridden by the --extern-html-root-url flag. This may be a good or bad thing depending on your perspective (should the user have control, or should the crate author?).

Maybe the flag addresses this, too? The user can override `-Z rustdoc-map` with `html_root_url` by passing the new flag (bikeshed: `--html-root-takes-precedence`). Or alternatively rustdoc can switch the default, and you can opt-in to having `--extern-html-root-url` take precedence, I don't feel strongly about what the default should be.

**Comment by KodrAus** [comment]
Kicked off an FCP in #229 for this.

**Comment by jyn514** [comment]
@KodrAus the FCP finished, so I think this can be merged :)

**Comment by SergioBenitez** [comment]
All: I am quite confused about the rationale to stop recommending this attribute. The crux of the argument, if I'm reading this right, is that 1) docs.rs overrides the URL when it builds docs anyway, and 2) it's hard to keep the URL up to date. I find the former to be rather problematic with consideration for decentralization, local docs, and development docs, and the latter a non-issue.

On 1: not only does this point entirely ignore local documentation building, but it also ignores crates that host their own documentation. Rocket, for example, hosts its docs at https://api.rocket.rs. Hosting our own docs allows us to provide redundancy, ensuring docs are accessible should something go wrong with docs.rs, while also allowing us to ensure docs are always up to date without waiting or relying on a third-party source. This latter point is especially important when releasing a new version. The delay between publishing a new release and docs.rs building and updating docs can sometimes be considerable. 

What's more, Rocket hosts docs for development versions that docs.rs does not and cannot build. See, for instance, https://api.rocket.rs/master. Without something like html_root_url or considerable manual effort to determine and set where every dependency's docs live, these docs, of at least equal importance, will contain broken links and be out of parity with the docs at docs.rs. It should be possible to independently build documentation that is as good as what's on docs.rs; removing the recommendation makes this considerably more difficult.

On 2: crates generally mention the version string in more places than just Cargo.toml, all of which need to be kept in sync as well, and so html_root_url shouldn't pose considerably more burden. For example, a crate's README often contains the TOML needed to add the crate as a dependency, which contains the version string. While this duplication is unfortunate, tools, like the one posted above, can ameliorate the situation. Rocket, for instance, automates changing the version number via a script to ensure they are all synchronized. Thus, adding one more version string to keep up-to-date should not pose considerably more burden, and certainly less so then needing to track down and keep up-to-date every dependency's preferred documentation URL should html_root_url be considered out of vogue entirely.

In summary, we should advocate for being able to build docs as good as those on docs.rs without being docs.rs for a variety of reasons, and the overhead to allowing this is minimal given existing realities. As such, it is my opinion that the recommendation for html_root_url should stay.

**Comment by jyn514** [comment]
@SergioBenitez you can still do this perfectly well by simply not passing `--no-deps` when you document your crate.

**Comment by jyn514** [comment]
Again, from https://github.com/rust-lang/api-guidelines/issues/229, the *only* case where `html_root_url` actually affects things is when a downstream crate is documented `--no-deps` and re-exports or links to one of your types. If any of those things aren't the case, then `html_root_url` does nothing at all.

**Comment by SergioBenitez** [comment]
@jyn514 This doesn't change any of the points I've made. I still cannot build docs like docs.rs can without being docs.rs. What's more, building the docs for all dependencies is wasteful, needlessly expensive, and creates clutter in the index.

**Comment by jyn514** [comment]
> I still cannot build docs like docs.rs can without being docs.rs.

You can build docs like docs.rs by passing `-Z rustdoc-map` the way it does: https://github.com/rust-lang/cargo/issues/8296. That does require a nightly toolchain, but it's significantly more reliable than `html_root_url`, which most crates *already* don't set.

If you want to go even further, you can build things exactly the same way as docs.rs by using the library we provide: https://github.com/rust-lang/docs.rs/blob/master/crates/metadata/lib.rs#L3.

I'm not trying to make a value judgement on whether or not you should use docs.rs. I just think this feature doesn't pull its weight, and I think it would better to stabilize `-Z rustdoc-map` instead of adding another task for library authors.

**Comment by SergioBenitez** [comment]
> You can build docs like docs.rs by passing `-Z rustdoc-map` the way it does: [rust-lang/cargo#8296](https://github.com/rust-lang/cargo/issues/8296). That does require a nightly toolchain, but it's significantly more reliable than `html_root_url`, which most crates _already_ don't set. If you want to go even further, you can build things exactly the same way as docs.rs by using the library we provide: https://github.com/rust-lang/docs.rs/blob/master/crates/metadata/lib.rs#L3.

Thank you for bringing `rustdoc-map` to my attention! I was not aware of it prior. This appears to solve the "considerable manual effort to determine and set where every dependency's docs live" _as long as_ we want to point all crate docs to a centralized source like docs.rs. It still does not provide a mechanism, however, for a crate author to dictate where its docs live and for its dependencies to point to that location without considerable effort. So while I see `rustdoc-map` as a wonderful fallback when `html_root_url` is not set, it is not a replacement.

> I'm not trying to make a value judgement on whether or not you should use docs.rs. I just think this feature doesn't pull its weight, and I think it would better to stabilize `-Z rustdoc-map` instead of adding another task for library authors.

I think a nice compromise would be to:
  * Set `rustdoc-map = docs.rs` by default.
  * Have cargo-doc determine a crate's doc locations via `html_root_url`, then `extern-html-root-url`, then `html_root_url`, in that order of preference.
  * Change the guidelines to suggest setting `html_root_url` _only when_ the preferred doc location is not `docs.rs`.

This would resolve all of my concerns while also avoiding the need for `html_root_url` when a crate author wants `docs.rs` to host their docs.

**Comment by jyn514** [comment]
> Have cargo-doc determine a crate's doc locations via html_root_url, then extern-html-root-url, then html_root_url, in that order of preference.

You said `html_root_url` twice - was there meant to be a third option?

**Comment by SergioBenitez** [comment]
> > Have cargo-doc determine a crate's doc locations via html_root_url, then extern-html-root-url, then html_root_url, in that order of preference.
> 
> 
> 
> You said `html_root_url` twice - was there meant to be a third option?

Indeed! Good catch. I meant to write `rustdoc-map` as the third option.

**Comment by jyn514** [comment]
@SergioBenitez rustdoc-map is implemented in terms of `extern-html-root-url`, they can't have different precedence.

**Comment by jyn514** [comment]
@KodrAus could you merge this? The FCP finished several months ago.
